No access control on OPTIONS method

stuartmcfarlane opened the pull request:
The OPTIONS method is neither a read nor write
operation. When used for preflight CORS header
discovery no auth can be sent.

This solution changed the isWrite method to
assert that an OPTIONS request is not a write
request. Further the ConfigurationBasedAccessController
is changed to allow requests are neither read
nor write.

stuartmcfarlane committed:
No access control on OPTIONS method

The OPTIONS method is neither a read nor write
operation. When used for preflight CORS header
discovery no auth can be sent.

This solution changed the isWrite method to
assert that an OPTIONS request is not a write
request. Further the ConfigurationBasedAccessController
is changed to allow requests are neither read
nor write.

cundd commented:
Thank you for the investigation and your work!

Does it finally solve your problem? Doesn't it result in a 404 response, or did you create a custom handler which implements a route for `OPTIONS`?

stuartmcfarlane commented:
I did add a n OPTIONS method handler on the route.

A more general solution might useful. Like all registered routes get an automatic OPTION method handler that is informed by the plugin.tx_rest.settings.responseHeaders typoscript settings. But the first step is disbling auth checks on the OPTIONS method

cundd commented:
Thank you for the clarification. I think I would prefer a separate method in addition to `isRead` and `isWrite`. Maybe called `isPreflight` or just `isOption`.

stuartmcfarlane commented:
perhaps requiresAuth() ?

cundd commented:
I would add the method to the Rest Request - which doesn't know if auth is required later in the pipeline

cundd commented:
I'm thinking if it would make sense to create a new `AccessControllerInterface::ACCESS` constant for such requests

stuartmcfarlane committed:
refactored isWrite

Removed isWrite from ConfigurationBasedAccessController.php
because it is only used here and does nothing but indirect to
request->isWrite

stuartmcfarlane committed:
refactored double code

`getAccess` and `requestNeedsAuthentication` were more or less identical

`getAccess` remains unchanged and `requestNeedsAuthentication` calls it
then maps the answer to true or false

stuartmcfarlane committed:
Added ACCESS_NOT_REQUIRED array of methods

stuartmcfarlane commented:
Adding it to the `Request` feels a bit odd. The `Request` knows nothing of authorization

cundd commented:
Oh. That's what I wanted to say. 

I'll add a `isOptions` or `isPreflight` to the request, afterwards. 

In the afternoon I'll try to fix the CI tests and then have a final look at your PR.

stuartmcfarlane commented:
I couldn't resist a little bit of refactoring. I've put it in separate commits. I'm happy enough to revert them if you don't like then.

cundd reviewed (changes requested):
I like the simplification of the code! But there is a small problem (if I didn't made a mistake). 

The line [ ConfigurationBasedAccessController.php#L161](https://github.com/stuartmcfarlane/rest/blob/54bc55e213e5741f6bc7911070a3e3681e11c5fb/Classes/Access/ConfigurationBasedAccessController.php#L161) will never evaluate to `true` since `getAccess()` will never return `AccessControllerInterface::ACCESS_REQUIRE_LOGIN` because `checkAuthentication()` will be called before. Do you think so too?

I think this is the result of an architecture failure of me.

stuartmcfarlane commented:
I found the code confusing and I believe immediately that I got it wrong 😄 . I'll have another look

cundd commented:
I'd create an additional method like the following and call that from the other two

```php
 private function getAccessConfiguration(RestRequestInterface $request)
    {
        if (!$this->requiresAuthorization($request)) {
            return AccessControllerInterface::ACCESS_ALLOW;
        }

        $configurationKey = self::ACCESS_METHOD_READ;
        if ($request->isWrite()) {
            $configurationKey = self::ACCESS_METHOD_WRITE;
        }

        $configuration = $this->getConfigurationForResourceType(new ResourceType($request->getResourceType()));
        // Throw an exception if the configuration is not complete
        if (!isset($configuration[$configurationKey])) {
            throw new InvalidConfigurationException($configurationKey . ' configuration not set', 1376826223);
        }

        return $configuration[$configurationKey];
    }
```

cundd commented:
I want to merge the changes. I'll merge and revert ddfb5be for now, so that we can work on it later :)
